### PR TITLE
feat: use error constants (#5088)

### PR DESCRIPTION
**Pull request by souravdasslg**
Refactor code to only throw errors defined in `constants/error-messages`.

#5088

### CODE REVIEW

**Comment by souravdasslg** [comment]
I tried to further categorize the error message types by adding prefixes. Any suggestions on that to make it more clear?

**Comment by rarkins** [comment]
This is great. I’m wondering if there’s a benefit to standardising prefixes now, e.g. `SYSTEM_`, `PLATFORM_`

File: lib/constants/error-messages.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+// Unknown Error
+export const UNKNOWN_ERROR = 'unknown-error';
+
+// File Error
+export const FILE_UPDATE_FAILED = 'update-failure';
+
+export const BUNDLER_INVALID_CREDENTIALS = 'bundler-credentials';
+export const BUNDLER_COULD_NOT_RESOLVE = 'bundler-resolve';
+export const BUNDLER_UNKNOWN_ERROR = 'bundler-unknown';
```
Comment (rarkins) [inline lib/constants/error-messages.ts]: If these three are "local" and only used within lib/manager/bundler/* then we don't need to define them here.

Comment (rarkins) [reply lib/constants/error-messages.ts]: Actually I just noticed we use these within branch worker: `} else if (err.message.startsWith('bundler-')) {`

**Comment by souravdasslg** [comment]
> This is great. I’m wondering if there’s a benefit to standardising prefixes now, e.g. `SYSTEM_`, `PLATFORM_`

There is couple of ambiguity is there, and possibly there will be a need to differentiate between similar errors like e.g `not-found` is the common error right now, either its about PR or Repository. Maybe there is a requirement to differentiate between these two errors in the future. This approach might be helpful by then.

**Comment by rarkins** [comment]
It's ok to differentiate between where the errors were thrown. Especially if the current approach is confusing.

File: lib/datasource/cargo/index.ts [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import { logger } from '../../logger';
 import got from '../../util/got';
 import { PkgReleaseConfig, ReleaseResult, Release } from '../common';
+import * as errorTypes from '../../constants/error-messages';
```
Comment (ViceIce) [inline lib/datasource/cargo/index.ts]:
```suggestion
import { REGISTRY_FAILURE } from '../../constants/error-messages';
```

I would prefer to direct import the specific errors.
@rarkins @JamieMagee What do you think?

File: lib/datasource/cargo/index.ts [hunk 3]
```diff
@@ -101,7 +102,7 @@ export async function getPkgReleases({
       (err.statusCode >= 500 && err.statusCode < 600)
     ) {
       logger.warn({ lookupName, err }, `cargo crates.io registry failure`);
-      throw new Error('registry-failure');
+      throw new Error(errorTypes.REGISTRY_FAILURE);
```
Comment (ViceIce) [inline lib/datasource/cargo/index.ts]:
```suggestion
      throw new Error(REGISTRY_FAILURE);
```

File: lib/datasource/cargo/index.ts (hunk 2 above)
Comment (rarkins) [reply lib/datasource/cargo/index.ts]: Yes, agreed. @souravdasslg could you make the changes?

**Review by viceice — CHANGES_REQUESTED** [review]
Please use direct imports as suggested

File: lib/datasource/cargo/index.ts (hunk 2 above)
Comment (souravdasslg) [reply lib/datasource/cargo/index.ts]: Sure.

File: lib/constants/error-messages.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
+// Platform Error
+export const PLATFORM_FAILURE = 'platform-failure';
+export const PLATFORM_NOT_FOUND = 'platform-not-found';
+export const BAD_CREDENTIALS = 'bad-credentials';
+export const INTEGRATION_UNAUTHORIZED = 'integration-unauthorized';
+export const RATE_LIMIT_EXCEEDED = 'rate-limit-exceeded';
+export const AUTHENTICATION_ERROR = 'authentication-error';
+export const AUTHENTICATION_FAILURE = 'authentication-failure';
```
Comment (rarkins) [inline lib/constants/error-messages.ts]: Looks like we should combine/consolidate/deduplicate these two. I think `AUTHENTICATION_ERROR` is the best to keep.

File: lib/constants/error-messages.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
+// Platform Error
```
Comment (rarkins) [inline lib/constants/error-messages.ts]: Rename the below to have `PLATFORM_` prefix

File: lib/constants/error-messages.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+export const PLATFORM_FAILURE = 'platform-failure';
+export const PLATFORM_NOT_FOUND = 'platform-not-found';
+export const BAD_CREDENTIALS = 'bad-credentials';
+export const INTEGRATION_UNAUTHORIZED = 'integration-unauthorized';
+export const RATE_LIMIT_EXCEEDED = 'rate-limit-exceeded';
+export const AUTHENTICATION_ERROR = 'authentication-error';
+export const AUTHENTICATION_FAILURE = 'authentication-failure';
+
+// System error
```
Comment (rarkins) [inline lib/constants/error-messages.ts]: Use `SYSTEM_` prefix and move to the top

File: lib/constants/error-messages.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+export const REPOSITORY_DISABLED = 'disabled';
+export const REPOSITORY_NOT_FOUND = 'not-found';
+export const REPOSITORY_ACCESS_FORBIDDEN = 'forbidden';
+export const REPOSITORY_BLOCKED = 'blocked';
+export const REPOSITORY_MIRRORED = 'mirror';
+export const REPOSITORY_FORKED = 'fork';
+export const REPOSITORY_NO_VULNERABILITY = 'no-vulnerability-alerts';
+export const REPOSITORY_UNINITIATED = 'uninitiated';
+
+// Manager Error
```
Comment (rarkins) [inline lib/constants/error-messages.ts]: Use `MANAGER_` prefix

File: lib/constants/error-messages.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+export const LOCKFILE_ERROR = 'lockfile-error';
+export const NO_PACKAGE_FILES = 'no-package-files';
+
+// Registry error
```
Comment (rarkins) [inline lib/constants/error-messages.ts]: Let's call these "datasource" errors instead, and use `DATASOURCE_` prefixes

**Review by rarkins — CHANGES_REQUESTED** [review]
@souravdasslg I think let's definitely take the opportunity to name these constants with some semantics, e.g. `SYSTEM_DISK_SPACE`, `PLATFORM_AUTHENTICATION_ERROR`, etc.

**Comment by souravdasslg** [comment]
> @souravdasslg I think let's definitely take the opportunity to name these constants with some semantics, e.g. `SYSTEM_DISK_SPACE`, `PLATFORM_AUTHENTICATION_ERROR`, etc.

Superb 😄 . 
Btw what are your thoughts on the errors which are not specific?
e.g, `throw new Error('invalid preset JSON')` or `throw new Error('Init: You must configure an Azure DevOps token');` etc ?

**Comment by rarkins** [comment]
> Btw what are your thoughts on the errors which are not specific?
> e.g, `throw new Error('invalid preset JSON')` or `throw new Error('Init: You must configure an Azure DevOps token');` etc ?

I think let's try to avoid them - ideally we should only throw errors defined in our constants from now on. If those long error messages are because we use it to log a warning or error (e.g. `log.warn(err.message)`), then let's log the message separately and then raise a standardized error based on our constants.

**Comment by souravdasslg** [comment]
> > Btw what are your thoughts on the errors which are not specific?
> > e.g, `throw new Error('invalid preset JSON')` or `throw new Error('Init: You must configure an Azure DevOps token');` etc ?
> 
> I think let's try to avoid them - ideally we should only throw errors defined in our constants from now on. If those long error messages are because we use it to log a warning or error (e.g. `log.warn(err.message)`), then let's log the message separately and then raise a standardized error based on our constants.

Then it is better to do this in separate PR I guess.

File: lib/constants/error-messages.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+export const REPOSITORY_MIRRORED = 'mirror';
+export const REPOSITORY_NOT_FOUND = 'not-found';
+export const REPOSITORY_NO_VULNERABILITY = 'no-vulnerability-alerts';
+export const REPOSITORY_RENAMED = 'renamed';
+export const REPOSITORY_TEMPORARY_ERROR = 'temporary-error';
+export const REPOSITORY_UNINITIATED = 'uninitiated';
+
+// Manager Error
+export const MANAGER_LOCKFILE_ERROR = 'lockfile-error';
+export const MANAGER_MANAGER_NO_PACKAGE_FILES = 'no-package-files';
```
Comment (rarkins) [inline lib/constants/error-messages.ts]: Single `MANAGER_` is enough here

Comment (souravdasslg) [reply lib/constants/error-messages.ts]: Silly of me. Fixing it.

File: lib/platform/bitbucket-server/index.ts [hunk 10]
```diff
@@ -1,4 +1,4 @@
-import url from 'url';
+import url, { URLSearchParams } from 'url';
```
Comment (ViceIce) [inline lib/platform/bitbucket-server/index.ts]: Why this?

**Review by viceice — COMMENTED** [review]
Hopefully you found all references.

File: lib/platform/bitbucket-server/index.ts (hunk 10 above)
Comment (souravdasslg) [reply lib/platform/bitbucket-server/index.ts]: I shouldn't have done this here. As this was not imported, I thought of adding that. Let me delete it for now.

https://github.com/renovatebot/renovate/blob/fc0b3d3026bd823c1ec81bc2c9c35a79aec1f6b6/lib/platform/bitbucket-server/index.ts#L360-L365

Comment (ViceIce) [reply lib/platform/bitbucket-server/index.ts]: I see, in older nodejs there was no global `URLSearchParams`

**Review by viceice — DISMISSED** [review]

**Comment by rarkins** [comment]
Before merging, I'd like to remove all instances literal constants from other files. e.g. when I search for `disk-space` in this branch then I see this:

image

I think the only instance of this string should be in lib/constants

**Review by rarkins — CHANGES_REQUESTED** [review]
Replace all string literals outside lib/constants

File: lib/platform/bitbucket-server/index.ts [hunk 11]
```diff
@@ -969,7 +975,9 @@ export async function updatePr(
   try {
     const pr = await getPr(prNo);
     if (!pr) {
-      throw Object.assign(new Error('not-found'), { statusCode: 404 });
+      throw Object.assign(new Error(REPOSITORY_NOT_FOUND), {
+        statusCode: 404,
```
[the remark this answers is not in the document]
Comment (ViceIce) [reply lib/platform/bitbucket-server/index.ts]: simply to assign `statusCode` to `Error`. Maybe we should add custom error classes with extended constructor to simplify these constructs.

```ts
export class HttpError extends Error {
  constructor(msg: string, private readonly statusCode: number) {
    super(msg);
  }
}

export class RenovateError extends Error {
  constructor(msg: string, props: Record<string,unknown>) {
    super(msg);
    Object.assign(this, props);
  }
}

throw new HttpError(REPOSITORY_NOT_FOUND, 404);
throw new RenovateError(REPOSITORY_NOT_FOUND, { statusCode: 404 });
```

[the remark this answers is not in the document]
Comment (rarkins) [reply lib/platform/bitbucket-server/index.ts]: Did we *use* this 404 anywhere though?

**Comment by souravdasslg** [comment]
> Before merging, I'd like to remove all instances literal constants from other files. e.g. when I search for `disk-space` in this branch then I see this:
> 
> image
> 
> I think the only instance of this string should be in lib/constants

Imagine a case, where we are matching an API response with 'system-failure' substring (`res.startsWith('system-failure')`), if we change this with error constant, then it's not reliable.
If someone changes the value of SYSTEM-FAILURE to something else, entire logic will fail.
In short, sometimes, the value of the constants also matters.
I avoided this as this is a much more significant change, and some of the error messages are common between API response and our error message.
Also, a couple of more test needs to be added if these changes are required.

File: lib/platform/bitbucket-server/index.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (ViceIce) [reply lib/platform/bitbucket-server/index.ts]: On a fast find: NOPE. so i think we can safely remove that

**Comment by rarkins** [comment]
I think we need to make these changes across the code base, but willing to ask @ViceIce as a second opinion. BTW I think that the `startsWith('disk-space')` example was an accidentally good one to pick because it's kind of semantically flawed once we move to constants. It likely can be changed to an exact match.

**Comment by ViceIce** [comment]
I think those `err.message.startsWith` cheks should only be happen on very low level catch blocks, where we catch error not thrown by renovate itself. All other errors should be checked with an exact match to the constant.

@souravdasslg You are right, this wouldn't be a simple refactoring. So maybe this pr should then be a feature? 🤔

File: lib/platform/bitbucket-server/index.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (souravdasslg) [reply lib/platform/bitbucket-server/index.ts]: I tried to change this and ended up breaking a couple of tests, so I skipped it for now.

**Review by rarkins — APPROVED** [review]

**Pull request merged**